autospec/infile_bb_parser.py

```diff
@@ -0,0 +1,202 @@
+        print('file has no version: {}')
+
+    return bb_dict
+
+
+def replace_pv(bb_dict):
+    for k, v in bb_dict.items():
+        if "${PV}" in v and "VERSION" in bb_dict:
+            v = v.replace("${PV}", bb_dict.get('VERSION'))
+        if "${ROS_SPN}" in v:
+            if "ROS_SPN" in bb_dict:
+                v = v.replace("${ROS_SPN}", bb_dict.get("ROS_SPN"))
+            elif "NAME" in bb_dict:
+                v = v.replace("${ROS_SPN}", bb_dict.get('NAME'))
+
+        bb_dict[k] = v
+
+
+def get_src_url(bb_dict):
+    if bb_dict.get('SRC_URI'):
+        bb_dict["URL"] = bb_dict.get("SRC_URI").split('\\')[0].split(';')[0]
+
+
+def clean_values(value):
+    # remove beginning and end quote
+    if value.startswith('"') and value.endswith('"'):
+        value = value[1:-1]
+
+    # remove line continuations
+    value = "".join(value.split("\\"))
+    return value
+
+
+def update_inherit(line, bb_dict):
+    if 'inherits' in bb_dict:
+        bb_dict['inherits'].append(' '.join(line.split(' ', 1)[1:]))
+    else:
+        bb_dict['inherits'] = line.split(' ', 1)[1:]
+
+
+def read_in_command(line, depth):
+    for c in line:
+        if c == '{':
+            depth += 1
+        if c == '}' and depth > 0:
+            depth -= 1
+        if c != "\n":
+            continue
+    return depth
+
+
+def pattern_match_line(line):
+
+    operators = ["??=", "?=", ":=", "+=",
+                 "=+", ".=", "=.", "="]
+
+    for i, e in enumerate(operators):
+        operators[i] = '\\' + '\\'.join(e)
+
+    # Split line to be [Key, operator, value] if in operators list
+    oper_pattern = r"(^[A-Z]+[_\-${}\[\]A-Za-z0-9]*)\s(" + '|'.join(
+        operators) + r")\s(\".*\")"
+
+    return re.compile(oper_pattern).search(line)
+
+
+def evaluate_expr(op, cur_val, assignment):
+    if not cur_val:
+        return assignment
+    return operators_dict.get(op).format(cur_val, assignment)
+
+
+def write_to_dict(bb_dict, m):
+
+    if len(m.groups()) == 3:
+        key = m.group(1)
+        value = clean_values(m.group(3))
+        op = m.group(2)
+
+        #
+        # ??= is the weakest variable assignment, so if that variable already
+        # has an assignment, do not overwrite it. = has the highest precedence
+        # and ?= is between the two.
+        #
+        if key in bb_dict and op == "??=":
+            return bb_dict
+        elif key in bb_dict and op == "?=":
+            if not isinstance(bb_dict[key], list):
+                return bb_dict
+
+        if key in bb_dict and isinstance(bb_dict[key], list):
+            v = bb_dict[key]
+            del [v[1]]
+            bb_dict[key] = "".join(v)
+        try:
+            bb_dict[key] = evaluate_expr(op, bb_dict.get(key), value)
+            if op == "??=":
+                bb_dict[key] = [bb_dict[key], 1]
+        except AttributeError as error:
+            print("Missing operation functionality: {}".format(error))
+
+    return bb_dict
+
+
+def bb_scraper(bb_fp, bb_dict):
```

Comment:
It would be much simpler to pass in a list of lines to this function instead of the open file handler. Then instead of calling `next()` in an infinite `while` loop you can iterate on the line list. This would simplify a lot of your code.

Reply:
I thought about this, but do you think the extra storing and double iteration would be costly?

Reply:
Not really, the bitbake recipes I've seen aren't that big. Besides, autospec already has to scan enormous build logs potentially several times per build so this would be a drop in the bucket. I'd rather have it easier to read and understand.

Reply:
Do you think the storing of the list, and extra iteration would be costly?

Reply:
err, see above.